pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -115,186 +123,302 @@ public boolean isDuplicate(MessageId messageId) {
-        if (acknowledgementGroupTimeMicros == 0 || !properties.isEmpty()) {
-            doImmediateBatchIndexAck(msgId, batchIndex, batchSize, ackType, properties,
-                    txn == null ? -1 : txn.getTxnIdMostBits(),
-                    txn == null ? -1 : txn.getTxnIdLeastBits());
-        } else if (ackType == AckType.Cumulative) {
-            BitSetRecyclable bitSet = BitSetRecyclable.create();
-            bitSet.set(0, batchSize);
-            bitSet.clear(0, batchIndex + 1);
-            doCumulativeAck(msgId, bitSet);
-        } else if (ackType == AckType.Individual) {
-            ConcurrentBitSetRecyclable bitSet;
-            if (txn != null) {
-                synchronized (txn) {
-                    ConcurrentHashMap<MessageIdImpl, ConcurrentBitSetRecyclable> transactionIndividualBatchIndexAcks =
-                            pendingIndividualTransactionBatchIndexAcks
-                                    .computeIfAbsent(txn, (v) -> new ConcurrentHashMap<>());
-                    bitSet = transactionIndividualBatchIndexAcks.computeIfAbsent(msgId, (v) -> {
-                        ConcurrentBitSetRecyclable value;
-                        value = ConcurrentBitSetRecyclable.create();
-                        value.set(0, msgId.getAcker().getBatchSize());
-                        return value;
-                    });
-                    bitSet.clear(batchIndex);
+    private MessageIdImpl modifyBatchMessageIdAndStatesInConsumer(BatchMessageIdImpl batchMessageId) {
+        MessageIdImpl messageId = new MessageIdImpl(batchMessageId.getLedgerId(),
+                batchMessageId.getEntryId(), batchMessageId.getPartitionIndex());
+        consumer.getStats().incrementNumAcksSent(batchMessageId.getBatchSize());
+        clearMessageIdFromUnAckTrackerAndDeadLetter(messageId);
+        return messageId;
+    }
+
+    private void modifyMessageIdStatesInConsumer(MessageIdImpl messageId) {
+        consumer.getStats().incrementNumAcksSent(1);
+        clearMessageIdFromUnAckTrackerAndDeadLetter(messageId);
+    }
+
+    private void clearMessageIdFromUnAckTrackerAndDeadLetter(MessageIdImpl messageId) {
+        consumer.getUnAckedMessageTracker().remove(messageId);
+        if (consumer.getPossibleSendToDeadLetterTopicMessages() != null) {
+            consumer.getPossibleSendToDeadLetterTopicMessages().remove(messageId);
+        }
+    }
+
+    private CompletableFuture<Void> doIndividualAck(MessageIdImpl messageId, Map<String, Long> properties) {
+        if (acknowledgementGroupTimeMicros == 0 || (properties != null && !properties.isEmpty())) {
+            // We cannot group acks if the delay is 0 or when there are properties attached to it. Fortunately that's an
+            // uncommon condition since it's only used for the compaction subscription.
+            return doImmediateAck(messageId, AckType.Individual, properties, null);
+        } else {
+            if (ackReceiptEnabled) {
```

Comment:
It's better to use a method isAckReceiptEnabled() to check if the consumer enabled the ack receipt and the broker support the ack receipt.